### ISSUE TITLE
Scala CustomResource does not conform to method customResources's type parameter bounds

### ISSUE THREAD

**Issue by ifilonenko** [fabric8io/kubernetes-client#1341]
EDIT: Works with scala 2.11 but not 2.12
I have been trying to create a CustomResource programatically using the JavaSDK via Scala. 

I implemented the following in Scala (mimicing the examples found here: https://github.com/fabric8io/kubernetes-client/tree/master/kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/crds) to implement the necessary classes to create the customResource.

```scala
import com.fasterxml.jackson.databind.JsonDeserializer
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import io.fabric8.kubernetes.api.builder.{Function => func}
import io.fabric8.kubernetes.api.model.{KubernetesResource, PodTemplateSpec}
import io.fabric8.kubernetes.client.{CustomResource, CustomResourceDoneable, CustomResourceList}

@JsonDeserialize(
  using = classOf[JsonDeserializer.None]
)
private[k8s] class ExecPodScaler(var spec: ExecPodScalerSpec) extends CustomResource {
  override def toString: String = {
    s"ExecPodScaler{" +
      s"apiVersion='$getApiVersion'\'" +
      s"kind='$getKind'\'" +
      s"metadata='$getMetadata'\'" +
      s"spec='$getSpec'}"
  }
  def getSpec: ExecPodScalerSpec = spec
  def setSpec(_spec: ExecPodScalerSpec): Unit = {
    spec = _spec
  }
}

@JsonDeserialize(
  using = classOf[JsonDeserializer.None]
)
private[k8s] class ExecPodScalerSpec(
  var numExecPods: Int,
  var podTemplates: Seq[PodTemplateSpec]) extends KubernetesResource[ExecPodScaler] {
  override def toString: String = {
    s"ExecPodScalerSpec{" +
      s"numExecPods='$getNumExecPods'\'" +
      s"podTemplates='$getPodTemplate'}"
  }

  def getNumExecPods: Int = numExecPods
  def setNumExecPods(_numExecPods: Int): Unit = {
    numExecPods = _numExecPods
  }

  def getPodTemplate: Seq[PodTemplateSpec] = podTemplates
  def setPodTemplates(_podTemplates: Seq[PodTemplateSpec]): Unit = {
    podTemplates = _podTemplates
  }
}

@JsonDeserialize(
  using = classOf[JsonDeserializer.None]
)
private[k8s] class ExecPodScalerList extends CustomResourceList[ExecPodScaler]

@JsonDeserialize(
  using = classOf[JsonDeserializer.None]
)
private[k8s] class DoneableExecPodScaler(
    val ecs: ExecPodScaler, val function: func[ExecPodScaler, ExecPodScaler])
  extends CustomResourceDoneable[ExecPodScaler](ecs, function) {
}
```
However, upon running: 
```scala
kubernetesClient.customResources[
      ExecPodScaler, ExecPodScalerList, DoneableExecPodScaler](
      execPodScalerCRD,
      classOf[ExecPodScaler],
      classOf[ExecPodScalerList],
      classOf[DoneableExecPodScaler]).inNamespace(namespace)
```
I get the following error: 
```
error: type arguments [ExecPodScaler,ExecPodScalerList,DoneableExecPodScaler]
 do not conform to method customResources's type parameter bounds
 [T <: io.fabric8.kubernetes.api.model.HasMetadata,
 L <: io.fabric8.kubernetes.api.model.KubernetesResourceList[
_ <: io.fabric8.kubernetes.api.model.HasMetadata],
 D <: io.fabric8.kubernetes.api.model.Doneable[T]]
    epsClient = kubernetesClient.customResources[
                                                ^
```
even tho by extending CustomResource, CustomResourceList, and CustomResourceDonable it should conform to type parameters. 

Thanks! Any help would be appreciated!!!

**Comment by erikerlandson** [comment]
A couple other notes on this:
1. @ifilonenko found that this compiled correctly on scala 2.11 - the problem shows up when compiling against 2.12
2. The signature that scala's error message is reporting is different than the signature implied by the java code for `customResources` method - from the scala side it looks as if scala is mis-interpreting the signature and that is what is causing the failure to resolve the method call.

**Comment by erikerlandson** [comment]
cc @skonto

**Comment by skonto** [comment]
@lrytz is this is a known issue? ^^^

**Comment by skonto** [comment]
@ifilonenko could you provide some context (execPodScalerCRD?

**Comment by ifilonenko** [comment]
It is just an existing CRD that I find using: 
```java
    public CustomResourceDefinition findCRD() {
        try {
            CustomResourceDefinitionList crds = client.customResourceDefinitions().list();
            List<CustomResourceDefinition> crdsItems = crds.getItems();
            logger.info("Number of crds: " + crdsItems.size());
            for ( CustomResourceDefinition crd: crdsItems ) {
                if (crd.getSpec().getGroup().toString().equals(EPS_CRD_GROUP) &&
                    crd.getSpec().getNames().getKind().toString().equals(EPS_KIND)) {
                    return crd;
                }
            }
        } catch (KubernetesClientException e) {
            logger.error("Problem with retrieving CRDs with Kubernetes Client", e);
        } catch (Exception e) {
            logger.error(e.getMessage(), e);
        }
        return null;
    }
```
I didn't realize the CRD definition itself mattered:
```yml
Name:         execpodscalers.<EPS_GROUP>
Namespace:
Labels:       <none>
Annotations:  <none>
API Version:  apiextensions.k8s.io/v1beta1
Kind:         CustomResourceDefinition
Metadata:
  Creation Timestamp:  2019-02-04T20:39:06Z
  Generation:          1
  Resource Version:    ....
  Self Link:           /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/execpodscalers.<EPS_GROUP>
  UID:                 ...
Spec:
  Group:  <EPS_GROUP>
  Names:
    Kind:       ExecPodScaler
    List Kind:  ExecPodScalerList
    Plural:     execpodscalers
    Singular:   execpodscaler
  Scope:        Namespaced
  Version:      v1
Status:
  Accepted Names:
    Kind:       ExecPodScaler
    List Kind:  ExecPodScalerList
    Plural:     execpodscalers
    Singular:   execpodscaler
  Conditions:
    Last Transition Time:  ...
    Message:               no conflicts found
    Reason:                NoConflicts
    Status:                True
    Type:                  NamesAccepted
    Last Transition Time:  ...
    Message:               the initial names have been accepted
    Reason:                InitialNamesAccepted
    Status:                True
    Type:                  Established
Events:                    <none>
```

**Comment by skonto** [comment]
@adriaanm I think this should be enough:

```
import io.fabric8.kubernetes.api.model.apiextensions.CustomResourceDefinition
val execPodScalerCRD = new CustomResourceDefinition
```

**Comment by skonto** [comment]
@ifilonenko I tried it form Spark repl, using @adriaanm code,
and created the CRD from here: https://v1-10.docs.kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/ on my minikube:

```
./bin/spark-shell --packages io.fabric8:openshift-client:4.1.0 
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
Spark context Web UI available at http://192.168.2.9:4040
Spark context available as 'sc' (master = local[*], app id = local-1549487638661).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.0.0-SNAPSHOT
      /_/
         
Using Scala version 2.12.8 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_171)
Type in expressions to have them evaluated.
Type :help for more information.

...
scala> res17.get(0)
res20: io.fabric8.kubernetes.api.model.apiextensions.CustomResourceDefinition = CustomResourceDefinition(apiVersion=apiextensions/v1beta1, kind=CustomResourceDefinition, metadata=ObjectMeta(annotations=null, clusterName=null, creationTimestamp=2019-02-06T21:11:18Z, deletionGracePeriodSeconds=null, deletionTimestamp=null, finalizers=[], generateName=null, generation=1, initializers=null, labels=null, name=crontabs.stable.example.com, namespace=null, ownerReferences=[], resourceVersion=404076, selfLink=/apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/crontabs.stable.example.com, uid=bed630e6-2a53-11e9-ba6f-a434d9270a65, additionalProperties={}), spec=CustomResourceDefinitionSpec(group=stable.example.com, names=CustomResourceDefinitionNames(kind=CronTa...

scala>   kubernetesClient.customResources[
     |         ExecPodScaler, ExecPodScalerList, DoneableExecPodScaler](
     |         res17.get(0),
     |         classOf[ExecPodScaler],
     |         classOf[ExecPodScalerList],
     |         classOf[DoneableExecPodScaler])
res21: io.fabric8.kubernetes.client.dsl.MixedOperation[ExecPodScaler,ExecPodScalerList,DoneableExecPodScaler,io.fabric8.kubernetes.client.dsl.Resource[ExecPodScaler,DoneableExecPodScaler]] = io.fabric8.kubernetes.client.dsl.internal.CustomResourceOperationsImpl@26feae9d

```
